hw/xbox/dsp/dsp_cpu.c

```diff
@@ -420,23 +431,29 @@ void dsp56k_reset_cpu(dsp_core_t* dsp)
     dsp->disasm_prev_inst_pc = 0xFFFFFFFF;
 }
 
-static OpcodeEntry lookup_opcode(uint32_t op) {
+static OpcodeEntry lookup_opcode_manually(uint32_t op)
+{
     OpcodeEntry r = {0};
     int i;
     for (i=0; i<ARRAYSIZE(nonparallel_opcodes); i++) {
         if ((op & nonparallel_matches[i][0]) == nonparallel_matches[i][1]) {
             if (nonparallel_opcodes[i].match_func 
                 && !nonparallel_opcodes[i].match_func(op)) continue;
             if (r.template != NULL) {
-                printf("qqq %x %s\n", op, r.template);
+                //printf("qqq %x %s\n", op, r.template);
             }
-            assert(r.template == NULL);
+            //assert(r.template == NULL);
```

Comment:
The way these are disabled is bad.

Also we still want to capture these cases, so you should return some meta-instruction handler which can `printf` and `assert` if necessary.

Reply:
could this be done in a ifdef debug like manner?

Reply:
Making assumptions about the idea behind these, you can keep them here (but enable them).

This check must also be for `(r != NULL) && (r.template == NULL)`, otherwise `r` might be unset.

As the printf is closely related to this, I suggest moving the `r != NULL` check into a separate condition to handle this debug stuff, maybe even both of them and an ugly `assert(false)`.

---

If all `OpcodeEntry`s have a template set (I'd assume so), you can reduce these checks to `r == NULL` / `r != NULL`.

In that case, `r.template == NULL` would have been a check if `r` was ever modified after zero-intialization (or: if the loop found more than one result). For pointers this just translates to `r == NULL`
Similar for the `printf` before: the assumption is that if `r.template != NULL` then there was a duplicate, so it should be printed, so it can be debugged.